python_scripts/02_numerical_pipeline_cross_validation.py

```diff
@@ -0,0 +1,151 @@
+# %% [markdown]
+# # Model evaluation using cross-validation
+#
+# In this notebook, we will still use only numerical features.
+#
+# We will discuss the practical aspects of assessing the generalization
+# performance of our model via **cross-validation** instead of a single
+# train-test split.
+# 
+# ## Data preparation
+#
+# First, let's load the full adult census dataset.
+
+# %%
+import pandas as pd
+
+adult_census = pd.read_csv("../datasets/adult-census.csv")
+
+# %% [markdown]
+# We will now drop the target from the data we will use to train our
+# predictive model.
+
+# %%
+target_name = "class"
+target = adult_census[target_name]
+data = adult_census.drop(columns=target_name)
+
+# %% [markdown]
+# Then, we select only the numerical columns, as seen in the previous
+# notebook.
+
+# %%
+numerical_columns = ["age", "capital-gain", "capital-loss", "hours-per-week"]
+
+data_numeric = data[numerical_columns]
+
+# %% [markdown]
+# Finally, we can now create a model using the `make_pipeline` tool to chain the
+# preprocessing and the estimator in every iteration of the cross-validation.
+
+# %%
+from sklearn.preprocessing import StandardScaler
+from sklearn.linear_model import LogisticRegression
+from sklearn.pipeline import make_pipeline
+
+model = make_pipeline(StandardScaler(), LogisticRegression())
+
+# %% [markdown]
+# ## The need for cross-validation
+#
+# In the previous notebook, we split the original data into a training set and a
+# testing set. But the score of a model will in general depend on the way we make
+# such a split. One downside of doing a single split is that it does not give
+# any information about the variability of the score. Another downside is that, in
+# a setting where the amount of data is small, the data available for training
+# and testing will be even smaller after splitting.
+#
+# Instead, we can use cross-validation. Cross-validation consists of repeating
+# the procedure such that the training and testing sets are different each
+# time. Generalization performance metrics are collected for each repetition and
+# then aggregated. As a result we can get an estimate of the variability of the
```

Comment:
The model (true) generalization performance is not variable since it the expected value of the score on the true distribution (which is unknown). What is variable is our measure (or estimate) of the test score be cause we work with random finite samples.

```suggestion
# then aggregated. As a result we can assess the variability of our measure of the
```